forecasting/prediction_intervals.py

```diff
@@ -0,0 +1,310 @@
+   
+    """
+    Inputs:
+        df: Historical time series data
+        horizon: Number of timesteps forecasted into the future
+        period: Seasonal period
+        bootstrap: toggle bootstrap or normal prediction interval
+        repetitions: Number of bootstrap repetitions
+        pred_width: width of prediction interval interval (an integer between 0 and 100)
+    Output:
+        pandas.DataFrame: a bootstrapped or normal prediction interval for df
+    """
+
+    if bootstrap:
+
+        ###Seasonal bootstrap###
+
+        #setting up df to pass into stl
+        seasonal_bs = df
+        stl = STL(seasonal_bs.iloc[:,0],period=period).fit()
+
+        #bootstrapping the trend for df
+        trend = stl.trend.to_frame()
+        trend_pi = naive_pi(trend,horizon,bootstrap=True, repetitions=repetitions, pred_width=pred_width)
+
+        #performing seasonal naive on the seasonal data
+        seasonal = stl.seasonal.to_frame()
+        seasonal_pi = s_naive_pi(seasonal, horizon,repetitions=repetitions, period=period)
+
+        return trend_pi + seasonal_pi
+        
+    else:
+
+        ###Seasonal naive###
+
+        #calculating the seasonal naive forecast for df
+        forecast_df = forecast_dates(df,horizon)
+        forecast_df['forecast'] = s_naive(df,period,horizon)
+
+        #extending the forecast backwards in time so we can calculate the residuals
+        season = df.iloc[-period:,0].to_list()
+        mult_season = int(len(df) / period) + 1
+        extended_forecast = (season * mult_season)[-len(df) % period: ]
+        sd_residuals = residual_sd(df,extended_forecast, no_missing_values=period)
+
+        #using the number of seasons prior to each point to calculate the standard deviation of forecasted points
+        seasons_in_forecast = [np.floor((h-1) / period) for h in range(1,horizon+1)] 
+        forecast_sd = [sd_residuals * np.sqrt(seasons_in_forecast[h-1]+1) for h in range(1,horizon+1)]
+
+        #creating an output forecast with the seasonal naive forecast and prediction interval
+        output_forecast = pi_output(forecast_df,horizon,forecast_sd, pred_width)
+        
+        return output_forecast
+
+
+
+def drift_pi(df:pd.DataFrame,horizon:int,pred_width=95.0) -> pd.DataFrame:
+
+    """
+    Inputs:
+        df: Historical time series observations
+        horizon: Number of timesteps forecasted into the future
+        period: Seasonal period
+    Outputs:
+        pandas.DataFrame: a normal prediction interval for df usuing the drift method
+    """
+
+    #setting up the forecast dataframe
+    forecast_df = forecast_dates(df,horizon)
+    forecast_df['forecast'] = drift_method(df,horizon)
+
+    #extending the forecast to test the forecast on observed data
+    latest_obs = df.iloc[-1,0]
+    first_obs = df.iloc[0,0]
+
+    slope = (latest_obs - first_obs) / len(df)
+
+    extended_forecast = [first_obs + slope * i for i in range(1, len(df) + 1)]
+
+    #calculating the residuals
```

Comment:
Potentiaily use bootstrapping again for drift: 
`def drift_method(df, horizon):
    """
    Inputs:
        df (pandas.Series): Historical time series observations (in order)
        horizon (int): Number of timesteps forecasted into the future
    Outputs:
        list: Forecasted time series
    """

    latest_obs = df.iloc[-1,0]
    first_obs = df.iloc[0,0]

    slope = (latest_obs - first_obs) / (len(df) - 1)

    forecast_list = [latest_obs + slope * h for h in range(1, horizon + 1)]

    return forecast_list

fitted_values = [np.nan, np.nan]
for x in range(2, len(df)):
    fcst = drift_method(df.set_index('ds')[:x], horizon=1)
    fitted_values.append(fcst[0])

df['slope'] = fitted_values

df['error'] = df['slope'] - df['y']

df.error.plot(kind='hist', bins=50)`